### PR TITLE
[Merged by Bors] - feat: add instances for intervals

### PR DESCRIPTION
**Pull request by kim-em**
Don't mind at all if anyone would like to push refactors or golfs. My main requirement from this PR is that
```lean
import Mathlib

example : WellFoundedLT { x : ℕ // x ≤ 37 }ᵒᵈ := inferInstance
```
works out of the box.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Intervals/Basic.lean [hunk 1]
```diff
@@ -1947,3 +1947,38 @@ instance : NoMaxOrder (Set.Iio x) :=
     exact ⟨⟨b, hb₂⟩, hb₁⟩⟩
 
 end Dense
+
+/-! ### Lemmas about intervals in subtypes
```
Comment (urkud) [inline Mathlib/Data/Set/Intervals/Basic.lean:1951]: IMHO, this file is already too large. What do you think about moving these lemmas to a new file?

File: Mathlib/Data/Set/Intervals/Basic.lean [hunk 2]
```diff
@@ -1947,3 +1947,38 @@ instance : NoMaxOrder (Set.Iio x) :=
...
+-/
+
+theorem Set.image_subtype_val_Icc_subset [Preorder α] {p : α → Prop} (a b : {x // p x}) :
```
Comment (urkud) [inline Mathlib/Data/Set/Intervals/Basic.lean:1954]: I would use `MapsTo Subtype.val (Icc a b) (Icc a.val b.val)`, then `(mapsTo_subtype_val_Icc _ _).image_subset` whenever you need the original statement. BTW, do we have these lemmas for any `Monotone`/`StrictMono` map?

File: Mathlib/Data/Set/Intervals/Basic.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Data/Set/Intervals/Basic.lean:1951]: I've moved to another file.

File: Mathlib/Data/Set/Intervals/Basic.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/Data/Set/Intervals/Basic.lean:1954]: I've added all the `mapsTo` lemmas requested, as well as `Monotone.image_Icc_subset` and friends, but I've kept the original lemmas too.

File: Mathlib/Order/LocallyFinite.lean [hunk 3]
```diff
@@ -1391,3 +1392,107 @@ theorem Set.finite_iff_bddBelow_bddAbove [Nonempty α] [Lattice α] [LocallyFini
...
+-- As the instances below are noncomputable,
+-- we make them low priority so when alternative constructions are available they are preferred.
+-- Someone may like to upgrade these instances to computable ones.
+
+noncomputable instance (priority := low) [Preorder α] [LocallyFiniteOrder α] :
+    LocallyFiniteOrder { x : α // x ≤ y } :=
+  LocallyFiniteOrder.ofFiniteIcc fun a b =>
+    Set.Finite.of_finite_image
+      (Set.Finite.subset (Set.finite_Icc ↑a ↑b) (Set.image_subtype_val_Icc_subset a b))
+      Set.injOn_subtype_val
```
Comment (eric-wieser) [inline Mathlib/Order/LocallyFinite.lean:1405]: Can't `Subtype.instLocallyFiniteOrder` already synthesize these?

Comment (semorrison) [reply Mathlib/Order/LocallyFinite.lean:1405]: Ugh, it does, assuming you have `open Classical`. I wish we didn't care about computability!

I've removed these.

File: Mathlib/Order/LocallyFinite.lean [hunk 4]
```diff
@@ -1391,3 +1392,80 @@ theorem Set.finite_iff_bddBelow_bddAbove [Nonempty α] [Lattice α] [LocallyFini
...
+  finsetIio a := Finset.Ico ⟨y, by rfl⟩ a
+  finsetIic a := Finset.Icc ⟨y, by rfl⟩ a
+  finset_mem_Iic a b := by
+    simp only [Finset.mem_Icc, and_iff_right_iff_imp]
+    exact fun _ => b.property
+  finset_mem_Iio a b := by
+    simp only [Finset.mem_Ico, and_iff_right_iff_imp]
+    exact fun _ => b.property
+
+noncomputable instance (priority := low) [Preorder α] [LocallyFiniteOrder α] :
```
Comment (eric-wieser) [inline Mathlib/Order/LocallyFinite.lean:1434]: Do we not have a `subtype` instance that works for these too?

**Comment by semorrison** [comment]
@eric-wieser, would you be able to fix the breakage from the renaming?

**Comment by eric-wieser** [comment]
Whoops, sorry!

File: Mathlib/Order/LocallyFinite.lean [hunk 5]
```diff
@@ -1391,3 +1395,78 @@ theorem Set.finite_iff_bddBelow_bddAbove [Nonempty α] [Lattice α] [LocallyFini
     fun ⟨⟨a,ha⟩,⟨b,hb⟩⟩ ↦ (Set.finite_Icc a b).subset (fun x hx ↦ ⟨ha hx,hb hx⟩ )⟩
 
 end Finite
+
+/-! We make the instance below low priority so when alternative constructions are available they are
```
Comment (eric-wieser) [inline Mathlib/Order/LocallyFinite.lean:1399]:
```suggestion
/-! We make the instances below low priority so when alternative constructions are available they are
```

**Review by eric-wieser — APPROVED** [review]
bors d+

THanks!

File: Mathlib/Order/LocallyFinite.lean [hunk 6]
```diff
@@ -1391,3 +1395,78 @@ theorem Set.finite_iff_bddBelow_bddAbove [Nonempty α] [Lattice α] [LocallyFini
     fun ⟨⟨a,ha⟩,⟨b,hb⟩⟩ ↦ (Set.finite_Icc a b).subset (fun x hx ↦ ⟨ha hx,hb hx⟩ )⟩
 
 end Finite
+
+/-! We make the instances below low priority so when alternative constructions are available they are
+preferred. -/
```
Comment (semorrison) [inline Mathlib/Order/LocallyFinite.lean:1400]:
```suggestion
/-! We make the instances below low priority
so when alternative constructions are available they are preferred. -/
```

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**